docker/docker_image_dest.go

```diff
@@ -596,6 +637,125 @@ func (d *dockerImageDestination) putOneSignature(url *url.URL, signature []byte)
 	}
 }
 
+func (d *dockerImageDestination) putSignaturesToCosignAttachments(ctx context.Context, signatures []signature.Cosign, manifestDigest digest.Digest) error {
+	if !d.c.useCosignAttachments {
+		return errors.New("writing Cosign attachments is disabled by configuration")
+	}
+
+	ociManifest, err := d.c.getCosignAttachmentManifest(ctx, d.ref, manifestDigest)
+	if err != nil {
+		return nil
+	}
+	var ociConfig imgspecv1.Image // Most fields empty by default
+	if ociManifest == nil {
+		ociManifest = manifest.OCI1FromComponents(imgspecv1.Descriptor{
+			MediaType: imgspecv1.MediaTypeImageConfig,
+			Digest:    "", // We will fill this in later.
+			Size:      0,
+		}, nil)
+	} else {
+		configBlob, err := d.c.getOCIDescriptorContents(ctx, d.ref, ociManifest.Config, iolimits.MaxConfigBodySize,
+			none.NoCache) // FIXME: Pass BlobInfoCache to signature accessors?)
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal(configBlob, &ociConfig); err != nil {
+			return fmt.Errorf("parsing Cosign config %s in %s: %w", ociManifest.Config.Digest.String(),
+				d.ref.ref.Name(), err)
+		}
+	}
+
+	for _, sig := range signatures {
+		found := false
+		for _, layer := range ociManifest.Layers {
+			if cosignSignatureMatchesLayer(sig, layer) {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+
+		sigDesc, err := d.putBlobBytesAsOCI(ctx, sig.UntrustedPayload, sig.UntrustedMIMEType, private.PutBlobOptions{
+			Cache:      none.NoCache, // FIXME: Pass BlobInfoCache to signature accessors? Or is this just pointless noise??
+			IsConfig:   false,
+			EmptyLayer: false,
+			LayerIndex: nil,
+		})
+		if err != nil {
+			return err
+		}
+		sigDesc.Annotations = sig.UntrustedAnnotations
+		ociManifest.Layers = append(ociManifest.Layers, sigDesc)
+		ociConfig.RootFS.DiffIDs = append(ociConfig.RootFS.DiffIDs, sigDesc.Digest)
```

Comment:
Are the diff IDs needed?  I couldn't find something in the [cosign spec](https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md) but recall that cosign just ships an empty config.

Reply:
I don’t know.

It’s _quite_ possible they aren’t needed. I also thought cosign ships a completely empty config, but when looking what to write into that config, I found DiffID entries in a Cosign-generated file. That might be a `go-containerregistry` artifact, the API has concepts like “add a layer” which adds it both to the manifest and to DiffIDs.

At least distribution/distribution seems not to care, _except if the client is schema1-only_.

Historically Quay has been the registry with the strictest validation we’ve been hearing about, I have no idea what they do.

Reply:
Thanks for checking!

I also suspect `go-containerregistry` to be the source.  I am totally fine to add the DiffIDs.

Reply:
It’s a good thing to point out and to track — if we don’t need to add DiffIDs, we would eliminate a blob write for the config on every copy that adds signatures to an existing image with signatures.